File: plugins/wise/engine/src/steps/gate.ts
// `approval` and `ask` steps (gates, D13): build the Gate the harness renders, and turn the
// client's `answer` value into the step's terminal state.

import { RPC_INVALID_PARAMS } from "../protocol.ts";
import { RpcError } from "../rpc.ts";
import type { ApprovalStep, AskStep, Gate } from "../types.ts";
import { headline } from "./agent.ts";

export type GateStep = ApprovalStep | AskStep;

export const APPROVAL_OPTIONS: readonly { value: string; label: string }[] = [
  { value: "approve", label: "Approve" },
  { value: "reject", label: "Reject" },
];

export function isGateStep(step: { type: string }): step is GateStep {
  return step.type === "approval" || step.type === "ask";
}

/** Gate for a rendered step. Approval always offers approve / reject; ask mirrors its options. */
export function buildGate(step: GateStep, gateId: string): Gate {
  const gate: Gate = {
    gate_id: gateId,
    step: step.id,
    kind: step.type,
    message: step.message.trim(),
  };
  if (step.type === "approval") {
    gate.options = APPROVAL_OPTIONS.map((o) => ({ ...o }));
    return gate;
  }
  if (step.options && step.options.length > 0) {
    gate.options = step.options.map((o) => ({ value: o, label: o }));
  }
  if (step.allow_text !== undefined) gate.allow_text = step.allow_text;
  return gate;
}

export type GateDecision = {
  status: "completed" | "failed";
  verdict: string;
  /** `ask` only: the run output to record. */
  output?: { name: string; value: string };
};

function asText(value: string | string[]): string {
  return Array.isArray(value) ? value.join(", ") : value;
}

/** Validate and apply an answer. Throws `RpcError(INVALID_PARAMS)` on a value the gate refuses. */
export function decideGate(step: GateStep, value: string | string[]): GateDecision {
  const text = asText(value).trim();
  if (step.type === "approval") {
    if (text === "approve") return { status: "completed", verdict: "approved" };
    if (text === "reject") return { status: "failed", verdict: "rejected" };
    throw new RpcError(
      RPC_INVALID_PARAMS,
      `answer: approval ${step.id} takes "approve" or "reject", got ${JSON.stringify(text)}`,
    );
  }
  const options = step.options ?? [];
  const allowText = step.allow_text ?? options.length === 0;
  if (text === "" || (!allowText && !options.includes(text))) {
    throw new RpcError(
      RPC_INVALID_PARAMS,
      `answer: ask ${step.id} takes one of ${JSON.stringify(options)}, got ${JSON.stringify(text)}`,
    );
  }
  const name = step.output ?? step.id;
  return {
    status: "completed",
    verdict: headline(`${name}=${text}`),
    output: { name, value: text },
  };
}

File: plugins/wise/engine/wise_engine/steps/gate.py
from __future__ import annotations

from typing import Any

from ..render import _json
from ..rpc import RpcError
from ..scheduler import JS_WHITESPACE
from .agent import headline

APPROVAL_OPTIONS = (
    {"value": "approve", "label": "Approve"},
    {"value": "reject", "label": "Reject"},
)
Json = dict[str, Any]


def is_gate_step(step: Json) -> bool:
    return step["type"] in ("approval", "ask")


def build_gate(step: Json, gate_id: str) -> Json:
    gate = dict(
        gate_id=gate_id,
        step=step["id"],
        kind=step["type"],
        message=step["message"].strip(JS_WHITESPACE),
    )
    if step["type"] == "approval":
        gate["options"] = [dict(option) for option in APPROVAL_OPTIONS]
    else:
        if step.get("options"):
            gate["options"] = [dict(value=option, label=option) for option in step["options"]]
        if "allow_text" in step:
            gate["allow_text"] = step["allow_text"]
    return gate


def decide_gate(step: Json, value: str | list[str]) -> Json:
    text = (", ".join(value) if isinstance(value, list) else value).strip(JS_WHITESPACE)
    if step["type"] == "approval":
        if text == "approve":
            return dict(status="completed", verdict="approved")
        if text == "reject":
            return dict(status="failed", verdict="rejected")
        raise RpcError(
            -32602, f'answer: approval {step["id"]} takes "approve" or "reject", got {_json(text)}'
        )
    options = step.get("options", [])
    allow_text = step.get("allow_text", len(options) == 0)
    if text == "" or (not allow_text and text not in options):
        raise RpcError(
            -32602, f"answer: ask {step['id']} takes one of {_json(options)}, got {_json(text)}"
        )
    name = step.get("output", step["id"])
    return dict(
        status="completed", verdict=headline(f"{name}={text}"), output={"name": name, "value": text}
    )
